### PR TITLE
Fix derivatives for linalg.vector_norm(..., dtype=)

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #76551
* #76547

As per title

### CODE REVIEW

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 1]
```diff
@@ -300,14 +300,16 @@ Tensor norm_jvp(const Tensor& self_p, const Tensor& self_t, const optional<Scala
...
+  auto dtype = opt_dtype.value_or(self_p.scalar_type());
+  return norm_jvp(self_p.to(dtype), self_t.to(dtype), scalar_ord, norm, dim, keepdim).to(toRealValueType(dtype));
 }
 
-Tensor linalg_vector_norm_backward(Tensor grad, const Tensor& self, const Scalar& scalar_ord, Tensor norm, const at::OptionalIntArrayRef& opt_dim, bool keepdim) {
+Tensor linalg_vector_norm_backward(Tensor grad, const Tensor& self, const Scalar& scalar_ord, Tensor norm, const at::OptionalIntArrayRef& opt_dim, bool keepdim, const c10::optional<ScalarType> opt_dtype) {
   auto dim = opt_dim.value_or(IntArrayRef({}));
-  return norm_backward(grad, self, scalar_ord, norm, dim, keepdim);
+  auto dtype = opt_dtype.value_or(self.scalar_type());
+  return norm_backward(grad, self.to(dtype), scalar_ord, norm, dim, keepdim).to(self.scalar_type());
```
Comment (ngimel) [inline torch/csrc/autograd/FunctionsManual.cpp:312]: I don't think we need the last cast, autograd engine should handle type demotion itself (as it does e.g. for all type-promoting pointwise operations. As for casting self to dtype, I've checked norm_backward, and for most cases it's also not needed (it participates in pw operations with grad, so will be automatically promoted), except for non-special-cased p>1, which is rare enough.

Comment (Lezcano) [reply torch/csrc/autograd/FunctionsManual.cpp]: done!

**Comment by mruberry** [comment]
@albanD would you review this?

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 2]
```diff
@@ -301,14 +301,16 @@ Tensor norm_jvp(const Tensor& self_p, const Tensor& self_t, const optional<Scala
...
+  auto dtype = opt_dtype.value_or(self_p.scalar_type());
+  return norm_jvp(self_p.to(dtype), self_t.to(dtype), scalar_ord, norm, dim, keepdim).to(toRealValueType(dtype));
 }
 
-Tensor linalg_vector_norm_backward(Tensor grad, const Tensor& self, const Scalar& scalar_ord, Tensor norm, const at::OptionalIntArrayRef& opt_dim, bool keepdim) {
+Tensor linalg_vector_norm_backward(Tensor grad, const Tensor& self, const Scalar& scalar_ord, Tensor norm, const at::OptionalIntArrayRef& opt_dim, bool keepdim, const c10::optional<ScalarType> opt_dtype) {
   auto dim = opt_dim.value_or(IntArrayRef({}));
-  return norm_backward(grad, self, scalar_ord, norm, dim, keepdim);
+  auto dtype = opt_dtype.value_or(self.scalar_type());
+  return norm_backward(grad, self.to(dtype), scalar_ord, norm, dim, keepdim);
```
Comment (ngimel) [inline torch/csrc/autograd/FunctionsManual.cpp:313]: there's still a cast of `self.to(dtype)` here, is it needed? I hoped self would be automatically promoted due to pw operations with `grad`.

Comment (Lezcano) [reply torch/csrc/autograd/FunctionsManual.cpp]: Actually no explicit casts are needed! Everything just works.

**Review by albanD — APPROVED** [review]
SGTM !

**Comment by seemethere** [comment]
@pytorchbot revert this

Reverting the entire stack, errors originated from
* https://github.com/pytorch/pytorch/pull/76547

Failed internal builds due to ([Link for Meta Employees](https://www.internalfb.com/diff/D36494019?selected_signal=c2FuZGNhc3RsZV93b3JrZmxvd19ydW46MTgwMTQzOTg1MTUzNTQ3NzQ%3D&selected_signal_verification_phase=1&dst_version_fbid=1211273672948052)):
```
aten/src/ATen/native/LinearAlgebra.cpp:2496:9: error: unused type alias 'Int' [-Werror,-Wunused-local-typedef]
  using Int = IntArrayRef::value_type;
        ^
1 error generated.
Command failed with exit code 1.
```

**Comment by seemethere** [comment]
This is actually a false positive, thought this created the original merge conflict, will re-land

**Pull request closed without merging**